src/mod_roster_rdbms.erl

```diff
@@ -275,40 +246,86 @@ read_subscription_and_groups(LUser, LServer, LJID, GSFunc, GRFunc) ->
-                [], Groups).
+%% PgSQL returns CHAR as an integer
+%% MySQL returns CHAR as a string
+decode_subscription(<<X>>) -> decode_subscription(X);
+decode_subscription($B) -> both;
+decode_subscription($T) -> to;
+decode_subscription($F) -> from;
+decode_subscription($N) -> none.
+
+encode_subscription(both) -> <<"B">>;
+encode_subscription(to)   -> <<"T">>;
+encode_subscription(from) -> <<"F">>;
+encode_subscription(none) -> <<"N">>.
+
+decode_ask(<<X>>) -> decode_ask(X);
+decode_ask($S) -> subscribe;
+decode_ask($U) -> unsubscribe;
+decode_ask($B) -> both;
+decode_ask($O) -> out;
+decode_ask($I) -> in;
+decode_ask($N) -> none.
+
+encode_ask(subscribe) -> <<"S">>;
+encode_ask(unsubscribe) -> <<"U">>;
+encode_ask(both) -> <<"B">>;
+encode_ask(out)  -> <<"O">>;
+encode_ask(in)   -> <<"I">>;
+encode_ask(none) -> <<"N">>.
+
+record_to_row(#roster{us = {LUser, _Server},
+                      jid = JID, name = Nick, subscription = Subscription,
+                      ask = Ask, askmessage = AskMessage}) ->
+    BinJID = jid:to_binary(jid:to_lower(JID)),
+    ExtSubscription = encode_subscription(Subscription),
+    ExtAsk = encode_ask(Ask),
+    [LUser, BinJID, Nick, ExtSubscription, ExtAsk, AskMessage,
+     <<"N">>, <<>>, <<"item">>].
+
+groups_to_rows(#roster{us = {LUser, _LServer}, jid = JID, groups = Groups}) ->
+    BinJID = jid:to_binary(jid:to_lower(JID)),
+    lists:foldl(fun (<<>>, Acc) -> Acc;
+                    (Group, Acc) -> [[LUser, BinJID, Group] | Acc]
+                end, [], Groups).
+
+raw_to_record(LServer,
+              {User, SJID, Nick, ExtSubscription, ExtAsk, AskMessage,
+               _Server, _ExtSubscribe, _Type} = I) ->
+    case jid:from_binary(SJID) of
+        error ->
+            ?LOG_ERROR(#{what => roster_parse_failed, row => io_lib:format("~0p", [I])}),
+            error;
+        JID ->
+            LJID = jid:to_lower(JID),
+            Subscription = decode_subscription(ExtSubscription),
+            Ask = decode_ask(ExtAsk),
+            #roster{usj = {User, LServer, LJID},
+                    us = {User, LServer}, jid = LJID, name = Nick,
+                    subscription = Subscription, ask = Ask,
+                    askmessage = AskMessage}
+    end.
+
+decode_roster_rows(LServer, Rows, JIDGroups) ->
+    GroupsDict = pairs_to_dict(JIDGroups),
+    F = fun (Row) -> raw_to_record_with_group(LServer, Row, GroupsDict) end,
+    lists:flatmap(F, Rows).
+
+pairs_to_dict(Pairs) ->
+    F = fun ({K, V}, Acc) -> dict:append(K, V, Acc) end,
+    lists:foldl(F, dict:new(), Pairs).
+
+raw_to_record_with_group(LServer, I, GroupsDict) ->
+    case raw_to_record(LServer, I) of
+        %% Bad JID in database:
+        error -> [];
+        R ->
+            BinJID = jid:to_binary(R#roster.jid),
+            [R#roster{groups = dict_find_list(BinJID, GroupsDict)}]
+    end.
+
+dict_find_list(K, Dict) ->
+    case dict:find(K, Dict) of
+        {ok, Values} -> Values;
+        error -> []
+    end.
```

Comment:
I get the idea of splitting the function into smaller chunks, but this one looks a bit awkward on its own.
If we are trying to be generic here, I think a `dict_find(Key, Dict, Default)` would be best. Otherwise it's difficult to understand the `_list` suffix as the only thing it does is to set the default to an empty list.